firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/ModelFileDownloadService.java

```diff
@@ -0,0 +1,371 @@
+  synchronized Long scheduleModelDownload(@NonNull CustomModel customModel) {
+    if (downloadManager == null) {
+      return null;
+    }
+
+    if (customModel.getDownloadUrl() == null || customModel.getDownloadUrl().isEmpty()) {
+      return null;
+    }
+    // todo handle expired url here and figure out what to do about delayed downloads too..
+
+    // Schedule a new downloading
+    Request downloadRequest = new Request(Uri.parse(customModel.getDownloadUrl()));
+    // check Url is not expired - get new one if necessary...
+
+    // By setting the destination uri to null, the downloaded file will be stored in
+    // DownloadManager's purgeable cache. As a result, WRITE_EXTERNAL_STORAGE permission is not
+    // needed.
+    downloadRequest.setDestinationUri(null);
+    if (VERSION.SDK_INT >= VERSION_CODES.N) {
+      downloadRequest.setRequiresCharging(downloadConditions.isChargingRequired());
+      downloadRequest.setRequiresDeviceIdle(downloadConditions.isDeviceIdleRequired());
+    }
+
+    if (downloadConditions.isWifiRequired()) {
+      downloadRequest.setAllowedNetworkTypes(Request.NETWORK_WIFI);
+    }
+
+    long id = downloadManager.enqueue(downloadRequest);
+    // update the custom model to store the download id - do not lose current local file - in case
+    // this is a background update.
+    sharedPreferencesUtil.setDownloadingCustomModelDetails(
+        new CustomModel(
+            customModel.getName(),
+            customModel.getModelHash(),
+            customModel.getSize(),
+            id,
+            customModel.getLocalFilePath()));
+    return id;
+  }
+
+  @Nullable
+  @VisibleForTesting
+  synchronized Integer getDownloadingModelStatusCode(Long downloadingId) {
+    if (downloadManager == null || downloadingId == null) {
+      return null;
+    }
+
+    Integer statusCode = null;
+
+    try (Cursor cursor = downloadManager.query(new Query().setFilterById(downloadingId))) {
+
+      if (cursor != null && cursor.moveToFirst()) {
+        statusCode = cursor.getInt(cursor.getColumnIndex(DownloadManager.COLUMN_STATUS));
+      }
+
+      if (statusCode == null) {
+        return null;
+      }
+
+      if (statusCode != DownloadManager.STATUS_RUNNING
+          && statusCode != DownloadManager.STATUS_PAUSED
+          && statusCode != DownloadManager.STATUS_PENDING
+          && statusCode != DownloadManager.STATUS_SUCCESSFUL
+          && statusCode != DownloadManager.STATUS_FAILED) {
+        // Unknown status
+        statusCode = null;
+      }
+      return statusCode;
+    }
+  }
+
+  @Nullable
+  private synchronized ParcelFileDescriptor getDownloadedFile(Long downloadingId) {
+    if (downloadManager == null || downloadingId == null) {
+      return null;
+    }
+
+    ParcelFileDescriptor fileDescriptor = null;
+    try {
+      fileDescriptor = downloadManager.openDownloadedFile(downloadingId);
+    } catch (FileNotFoundException e) {
+      System.out.println("Downloaded file is not found");
```

Comment:
Log maybe?

Reply:
Yes - changing all these to logs in the next major PR. Just started working on it this morning.